### PR TITLE
[#4007] improvement: Use template to reduce Privileges duplicate codes

### PR DESCRIPTION
**Pull request by rich7420**
### What changes were proposed in this pull request?
- make the template to reduce these duplicate codes.

### Why are the changes needed?
- Currently [Privileges.java](https://github.com/datastrato/gravitino/blob/main/api/src/main/java/com/datastrato/gravitino/authorization/Privileges.java#L266) have more than one thousand duplicate code to instantiation CreateCatalog, AlterCatalog, ...
- we can use template to make Privileges.java smaller.
Fix: #4007

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

`./gradlew test`

### CODE REVIEW

**Comment by jerryshao** [comment]
@qqqttt123 @xunliu can you please help to review this?

**Comment by xunliu** [comment]
hi @rich7420 Thank you for your contributions.

Please add these two functions in the `public abstract static class GenericPrivilege<T extends GenericPrivilege<T>>`
```
    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      GenericPrivilege<?> that = (GenericPrivilege<?>) o;
      return condition == that.condition && name == that.name;
    }

    @Override
    public int hashCode() {
      return Objects.hash(condition, name);
    }
```

This way, it can fix the issue of `Assertions.assertEquals(metalake, anotherMetalake);`

**Comment by rich7420** [comment]
@xunliu ok! I will try this way. Thank you. I appreciate it.

**Comment by rich7420** [comment]
@xunliu Can I change the code like this? Because the original way can't compile successfully.
```
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        // Change here: replaced getClass() != o.getClass() with !(o instanceof GenericPrivilege)
        if (!(o instanceof GenericPrivilege)) return false;
        GenericPrivilege<?> that = (GenericPrivilege<?>) o;
        return condition == that.condition && name == that.name;
    }

    @Override
    public int hashCode() {
        return Objects.hash(condition, name);
    }
```

**Comment by rich7420** [comment]
Sorry , Could someone please help me rerun the test? The code that failed this time is the same as the code that succeeded before.

**Comment by qqqttt123** [comment]
Every privilege should return a singleton instance. Because we cache the role, the role will contain many privileges. If we use the singleton for every privilege. We can save much memory.

**Comment by rich7420** [comment]
@qqqttt123 According to the idea, Maybe I can change code like this to add `ALLOW_INSTANCE` and `DENY_INSTANCE`. 
What do you think?
```
  public static class CreateCatalog extends GenericPrivilege<CreateCatalog> {
    private static final CreateCatalog ALLOW_INSTANCE = new CreateCatalog(Condition.ALLOW, Name.CREATE_CATALOG);
    private static final CreateCatalog DENY_INSTANCE = new CreateCatalog(Condition.DENY, Name.CREATE_CATALOG);

    private CreateCatalog(Condition condition, Name name) {
      super(condition, name);
    }

    /** @return The instance with allow condition of the privilege. */
    public static CreateCatalog allow() {
      return ALLOW_INSTANCE;
    }

    /** @return The instance with deny condition of the privilege. */
    public static CreateCatalog deny() {
      return DENY_INSTANCE;
    }
  }
```

**Comment by qqqttt123** [comment]
> @qqqttt123 According to the idea, Maybe I can change code like this to add `ALLOW_INSTANCE` and `DENY_INSTANCE`. What do you think?
> 
> ```
>   public static class CreateCatalog extends GenericPrivilege<CreateCatalog> {
>     private static final CreateCatalog ALLOW_INSTANCE = new CreateCatalog(Condition.ALLOW, Name.CREATE_CATALOG);
>     private static final CreateCatalog DENY_INSTANCE = new CreateCatalog(Condition.DENY, Name.CREATE_CATALOG);
> 
>     private CreateCatalog(Condition condition, Name name) {
>       super(condition, name);
>     }
> 
>     /** @return The instance with allow condition of the privilege. */
>     public static CreateCatalog allow() {
>       return ALLOW_INSTANCE;
>     }
> 
>     /** @return The instance with deny condition of the privilege. */
>     public static CreateCatalog deny() {
>       return DENY_INSTANCE;
>     }
>   }
> ```

Ok for me.

**Review by xunliu — DISMISSED** [review]
hi @rich7420 Thank you for your contributions.

LGTM

**Comment by rich7420** [comment]
@xunliu thank you. I appreciate it!!!

**Comment by jerqi** [comment]
> > @qqqttt123 According to the idea, Maybe I can change code like this to add `ALLOW_INSTANCE` and `DENY_INSTANCE`. What do you think?
> > ```
> >   public static class CreateCatalog extends GenericPrivilege<CreateCatalog> {
> >     private static final CreateCatalog ALLOW_INSTANCE = new CreateCatalog(Condition.ALLOW, Name.CREATE_CATALOG);
> >     private static final CreateCatalog DENY_INSTANCE = new CreateCatalog(Condition.DENY, Name.CREATE_CATALOG);
> > 
> >     private CreateCatalog(Condition condition, Name name) {
> >       super(condition, name);
> >     }
> > 
> >     /** @return The instance with allow condition of the privilege. */
> >     public static CreateCatalog allow() {
> >       return ALLOW_INSTANCE;
> >     }
> > 
> >     /** @return The instance with deny condition of the privilege. */
> >     public static CreateCatalog deny() {
> >       return DENY_INSTANCE;
> >     }
> >   }
> > ```
> 
> Ok for me.

This comment seems not be addressed.

**Comment by rich7420** [comment]
> > > @qqqttt123 According to the idea, Maybe I can change code like this to add `ALLOW_INSTANCE` and `DENY_INSTANCE`. What do you think?
> > > ```
> > >   public static class CreateCatalog extends GenericPrivilege<CreateCatalog> {
> > >     private static final CreateCatalog ALLOW_INSTANCE = new CreateCatalog(Condition.ALLOW, Name.CREATE_CATALOG);
> > >     private static final CreateCatalog DENY_INSTANCE = new CreateCatalog(Condition.DENY, Name.CREATE_CATALOG);
> > > 
> > >     private CreateCatalog(Condition condition, Name name) {
> > >       super(condition, name);
> > >     }
> > > 
> > >     /** @return The instance with allow condition of the privilege. */
> > >     public static CreateCatalog allow() {
> > >       return ALLOW_INSTANCE;
> > >     }
> > > 
> > >     /** @return The instance with deny condition of the privilege. */
> > >     public static CreateCatalog deny() {
> > >       return DENY_INSTANCE;
> > >     }
> > >   }
> > > ```
> > 
> > 
> > Ok for me.
> 
> This comment seems not be addressed.

Thank you for your feedback. Added static constants ALLOW_INSTANCE and DENY_INSTANCE in each privilege class, These constants are initialized when the class is loaded, ensuring the instances are created only once.  What else can I improve on?

**Comment by jerqi** [comment]
> > > > @qqqttt123 According to the idea, Maybe I can change code like this to add `ALLOW_INSTANCE` and `DENY_INSTANCE`. What do you think?
> > > > ```
> > > >   public static class CreateCatalog extends GenericPrivilege<CreateCatalog> {
> > > >     private static final CreateCatalog ALLOW_INSTANCE = new CreateCatalog(Condition.ALLOW, Name.CREATE_CATALOG);
> > > >     private static final CreateCatalog DENY_INSTANCE = new CreateCatalog(Condition.DENY, Name.CREATE_CATALOG);
> > > > 
> > > >     private CreateCatalog(Condition condition, Name name) {
> > > >       super(condition, name);
> > > >     }
> > > > 
> > > >     /** @return The instance with allow condition of the privilege. */
> > > >     public static CreateCatalog allow() {
> > > >       return ALLOW_INSTANCE;
> > > >     }
> > > > 
> > > >     /** @return The instance with deny condition of the privilege. */
> > > >     public static CreateCatalog deny() {
> > > >       return DENY_INSTANCE;
> > > >     }
> > > >   }
> > > > ```
> > > 
> > > 
> > > Ok for me.
> > 
> > 
> > This comment seems not be addressed.
> 
> Thank you for your feedback. Added static constants ALLOW_INSTANCE and DENY_INSTANCE in each privilege class, These constants are initialized when the class is loaded, ensuring the instances are created only once. What else can I improve on?

You don't add static constants ALLOW_INSTANCE and DENY_INSTANCE in the code.

**Comment by rich7420** [comment]
@jerqi Sorry , I forgot to push. Thank you for your reply.

File: api/src/main/java/com/datastrato/gravitino/authorization/Privileges.java [hunk 1]
```diff
@@ -262,114 +229,148 @@ public static Privilege deny(Privilege.Name name) {
...
+
+    /**
+     * Creates an allowed privilege.
+     *
+     * @param <T> the type of the privilege
+     * @param name the name of the privilege
+     * @param factory the factory to create the privilege
+     * @return the created privilege instance
+     */
+    public static <T extends GenericPrivilege<T>> T allow(
```
Comment (qqqttt123) [inline api/src/main/java/com/datastrato/gravitino/authorization/Privileges.java:279]: Is this method used for other methods?

File: api/src/main/java/com/datastrato/gravitino/authorization/Privileges.java [hunk 2]
```diff
@@ -262,114 +229,148 @@ public static Privilege deny(Privilege.Name name) {
...
+
+    /**
+     * Creates a denied privilege.
+     *
+     * @param <T> the type of the privilege
+     * @param name the name of the privilege
+     * @param factory the factory to create the privilege
+     * @return the created privilege instance
+     */
+    public static <T extends GenericPrivilege<T>> T deny(
```
Comment (qqqttt123) [inline api/src/main/java/com/datastrato/gravitino/authorization/Privileges.java:292]: ditto.

**Comment by rich7420** [comment]
@qqqttt123 thank you for your reply. I removed that codes.

**Comment by xunliu** [comment]
hi @rich7420 Please fix CI issue.

```
BUILD FAILED in 58s
                -····private·static·final·DropCatalog·ALLOW_INSTANCE·=·new·DropCatalog(Condition.ALLOW,·Name.CREATE_CATALOG);
                -····private·static·final·DropCatalog·DENY_INSTANCE·=·new·DropCatalog(Condition.DENY,·Name.CREATE_CATALOG);
                +····private·static·final·DropCatalog·ALLOW_INSTANCE·=
                +········new·DropCatalog(Condition.ALLOW,·Name.CREATE_CATALOG);
                +····private·static·final·DropCatalog·DENY_INSTANCE·=
                +········new·DropCatalog(Condition.DENY,·Name.CREATE_CATALOG);
                 
                 ····private·DropCatalog(Condition·condition,·Name·name)·{
                 ······super(condition,·name);
                @@ -373,8 +379,11 @@
                 
                 ··/**·The·privilege·to·use·a·catalog.·*/
                 ··public·static·class·UseCatalog·extends·GenericPrivilege<UseCatalog>·{
                -····private·static·final·UseCatalog·ALLOW_INSTANCE·=·new·UseCatalog(Condition.ALLOW,·Name.CREATE_CATALOG);
                -····private·static·final·UseCatalog·DENY_INSTANCE·=·new·UseCatalog(Condition.DENY,·Name.CREATE_CATALOG);
                +····private·static·final·UseCatalog·ALLOW_INSTANCE·=
                +········new·UseCatalog(Condition.ALLOW,·Name.CREATE_CATALOG);
                +····private·static·final·UseCatalog·DENY_INSTANCE·=
                +········new·UseCatalog(Condition.DENY,·Name.CREATE_CATALOG);
            ... ([410](https://github.com/apache/gravitino/actions/runs/9778148018/job/26998543927?pr=4010#step:6:412) more lines that didn't fit)
        Run './gradlew :api:spotlessApply' to fix these violations.
```

Please run './gradlew :api:spotlessApply' to fix these violations.

**Comment by rich7420** [comment]
@xunliu thank you for your feedback.

**Review by xunliu — APPROVED** [review]
LGTM

**Pull request merged**